protocols/raft/src/main/java/io/atomix/protocols/raft/service/impl/DefaultServiceContext.java

```diff
@@ -360,6 +359,7 @@ private void maybeInstallSnapshot(long index) {
       // Persist the snapshot to disk in a background thread before completing the snapshot future.
       snapshotExecutor.execute(() -> {
         pendingSnapshot.persist();
+        pendingSnapshots.put(snapshotIndex, pendingSnapshot);
```

Comment:
@kuujo why not just persist if return value of put is true?

Reply:
```java
  snapshotExecutor.execute(() -> {
    if (pendingSnapshots.put(snapshotIndex, pendingSnapshot)) {
      pendingSnapshot.persist();
    }
    future.complete(snapshotIndex);
  });
```

Reply:
```java
  snapshotExecutor.execute(() -> {
    if (pendingSnapshots.put(snapshotIndex, pendingSnapshot)) {
      pendingSnapshot.persist();
    }
    future.complete(snapshotIndex);
  });
```

Reply:
No need. This part of the logic actually works fine, and this particular change really has no effect other than to add `PendingSnapshot`s to the map a bit later so they're not unnecessarily checked in `maybeCompleteSnapshot`. Even if we only want to `persist()` the snapshot if there's not already a snapshot at the same index in the `pendingSnapshots` map because a snapshot could still exist at the index from a previous run or could have already been persisted and removed from the `pendingSnapshots` map. The primary fix in this PR is to handle duplicate snapshots replicated from the leader. But perhaps some additional checks should also be added to avoid duplicate snapshots being taken of the state machine as well.

Reply:
The suggestion was mainly for aesthetic reasons, seems silly to invoke persist then add it to the pending map.